xds/src/main/java/io/grpc/xds/RingHashLoadBalancer.java

```diff
@@ -93,81 +86,147 @@ public boolean acceptResolvedAddresses(ResolvedAddresses resolvedAddresses) {
-    double normalizedMinWeight = (double) minWeight / totalWeight;
-    // Scale up the number of hashes per host such that the least-weighted host gets a whole
-    // number of hashes on the the ring. Other hosts might not end up with whole numbers, and
-    // that's fine (the ring-building algorithm can handle this). This preserves the original
-    // implementation's behavior: when weights aren't provided, all hosts should get an equal
-    // number of hashes. In the case where this number exceeds the max_ring_size, it's scaled
-    // back down to fit.
-    double scale = Math.min(
-        Math.ceil(normalizedMinWeight * config.minRingSize) / normalizedMinWeight,
-        (double) config.maxRingSize);
-    ring = buildRing(serverWeights, totalWeight, scale);
-
-    // Shut down subchannels for delisted addresses.
-    List<Subchannel> removedSubchannels = new ArrayList<>();
-    for (EquivalentAddressGroup addr : removedAddrs) {
-      removedSubchannels.add(subchannels.remove(addr));
+
+    return true;
+  }
+
+  /**
+   * Updates the overall balancing state by aggregating the connectivity states of all subchannels.
+   *
+   * <p>Aggregation rules (in order of dominance):
+   * <ol>
+   *   <li>If there is at least one subchannel in READY state, overall state is READY</li>
+   *   <li>If there are <em>2 or more</em> subchannels in TRANSIENT_FAILURE, overall state is
+   *   TRANSIENT_FAILURE</li>
+   *   <li>If there is at least one subchannel in CONNECTING state, overall state is
+   *   CONNECTING</li>
+   *   <li> If there is one subchannel in TRANSIENT_FAILURE state and there is
+   *    more than one subchannel, report CONNECTING </li>
+   *   <li>If there is at least one subchannel in IDLE state, overall state is IDLE</li>
+   *   <li>Otherwise, overall state is TRANSIENT_FAILURE</li>
+   * </ol>
+   */
+  @Override
+  protected void updateOverallBalancingState() {
+    checkState(!getChildLbStates().isEmpty(), "no subchannel has been created");
+    if (this.currentConnectivityState == SHUTDOWN) {
+      // Ignore changes that happen after shutdown is called
+      return;
     }
-    // If we need to proactively start connecting, iterate through all the subchannels, starting
-    // at a random position.
-    // Alternatively, we should better start at the same position.
-    connectionAttemptIterator = subchannels.values().iterator();
-    int randomAdvance = random.nextInt(subchannels.size());
-    while (randomAdvance-- > 0) {
-      connectionAttemptIterator.next();
+
+    // Calculate the current overall state considering sticky TF
+    int numIdle = 0;
+    int numReady = 0;
+    int numConnecting = 0;
+
+    forloop:
+    for (ChildLbState childLbState : getChildLbStates()) {
+      ConnectivityState state = childLbState.getCurrentState();
+      switch (state) {
+        case READY:
+          numReady++;
+          break forloop;
+        case CONNECTING:
+          numConnecting++;
+          break;
```

Comment:
This is not what I understand. And this contradicts with the second bullet point in the comment
> <li>If there are <em>2 or more</em> subchannels in TRANSIENT_FAILURE, overall state is TRANSIENT_FAILURE</li>

I think we still to count all states then conclude the overall state based on the comment algorithm. Then if the conclude state are not TF or first time TF, just report it. Otherwise do gRFC:
> if (in_transient_failure && endpoint_entered_transient_failure) {
>...

Reply:
The comment was the old algorithm which is no longer appropriate since we no longer treat the first 2 ring elements as special.  The `if` statement you posted was from the suggested picker logic, not from the overall state calculations.
I've updated the comment to reflect the new simple connection handling.

Reply:
I've restored the comment and updated the logic to match the comment (along with associated tests).